packages/@lwc/engine-dom/src/renderer.ts

```diff
@@ -43,39 +51,44 @@ function isCustomElementRegistryAvailable() {
 }
 
 if (isCustomElementRegistryAvailable()) {
-    getCustomElement = customElements.get.bind(customElements);
-    defineCustomElement = customElements.define.bind(customElements);
-    HTMLElementConstructor = HTMLElement;
-} else {
-    const registry: Record<string, CustomElementConstructor> = create(null);
-    const reverseRegistry: WeakMap<CustomElementConstructor, string> = new WeakMap();
-
-    defineCustomElement = function define(name: string, ctor: CustomElementConstructor) {
-        if (name !== StringToLowerCase.call(name) || registry[name]) {
-            throw new TypeError(`Invalid Registration`);
-        }
-        registry[name] = ctor;
-        reverseRegistry.set(ctor, name);
-    };
-
-    getCustomElement = function get(name: string): CustomElementConstructor | undefined {
-        return registry[name];
+    const { get, define } = customElements;
+    let iframeWindow: Window & typeof globalThis;
+    let cacheIframeWindow = () => {
+        cacheIframeWindow = () => iframeWindow;
+        // headless iframe used for local registry for LWC components
+        const localRegistryIframe = document.createElement('iframe');
+        localRegistryIframe.hidden = true;
+        (document.body || document).appendChild(localRegistryIframe);
+        iframeWindow = localRegistryIframe.contentWindow! as Window & typeof globalThis;
+        return iframeWindow;
     };
-
-    HTMLElementConstructor = function HTMLElement(this: HTMLElement) {
-        if (!(this instanceof HTMLElement)) {
-            throw new TypeError(`Invalid Invocation`);
-        }
-        const { constructor } = this;
-        const name = reverseRegistry.get(constructor as CustomElementConstructor);
-        if (!name) {
-            throw new TypeError(`Invalid Construction`);
+    getGlobalCustomElement = get.bind(customElements);
+    getUpgradableConstructor = (name: string) => {
+        class LWCUpgradableElement extends cacheIframeWindow().HTMLElement {
+            constructor(upgradeCallback?: UpgradeCallback) {
+                super();
+                // correcting the proto chain to avoid leaking internals of the iframe
+                setPrototypeOf(this, HTMLElement.prototype);
+                if (isFunction(upgradeCallback)) {
+                    upgradeCallback(this); // nothing to do with the result for now
+                }
+            }
         }
-        const elm = document.createElement(name);
-        setPrototypeOf(elm, constructor.prototype);
-        return elm;
+        define.call(cacheIframeWindow().customElements, name, LWCUpgradableElement);
```

Comment:
when registering the custom element inside the hidden iframe's registry, it must be a component that extends the hidden iframe's HTMLElement instead of the main window HTMLElement, that's an invariant.

Reply:
and that is fine, because we will later on correct the proto chain anyways, so, no leaking anything from inside the iframe anyways.